report_async/models/report_async.py

```diff
@@ -0,0 +1,181 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+
+import base64
+from odoo import api, fields, models, _
+from odoo.tools.safe_eval import safe_eval
+from odoo.exceptions import UserError
+from odoo.addons.queue_job.job import job
+
+
+# Define all supported report_type
+REPORT_TYPES_FUNC = {'qweb-pdf': 'render_qweb_pdf',
+                     'qweb-text': 'render_qweb_text',
+                     'qweb-xml': 'render_qweb_xml',
+                     'csv': 'render_csv',
+                     'excel': 'render_excel',
+                     'xlsx': 'render_xlsx', }
+
+
+class ReportAsync(models.Model):
+    _name = 'report.async'
+    _description = 'Report Async'
+
+    action_id = fields.Many2one(
+        comodel_name='ir.actions.act_window',
+        string='Reports',
+        required=True,
+    )
+    allow_async = fields.Boolean(
+        string='Allow Async',
+        default=False,
+        help="This is not automatic field, please check if you want to allow "
+        "this report in background process",
+    )
+    name = fields.Char(
+        string='Name',
+        related='action_id.display_name',
+    )
+    email_notify = fields.Boolean(
+        string='Email Notification',
+        help="Send email with link to report, when it is ready",
+    )
+    group_ids = fields.Many2many(
+        string='Groups',
+        comodel_name='res.groups',
+        help="Only user in selected groups can use this report."
+        "If left blank, everyone can use",
+    )
+    job_ids = fields.Many2many(
+        comodel_name='queue.job',
+        compute='_compute_job',
+        help="List all jobs related to this running report",
+    )
+    job_status = fields.Selection(
+        selection=[('pending', 'Pending'),
+                   ('enqueued', 'Enqueued'),
+                   ('started', 'Started'),
+                   ('done', 'Done'),
+                   ('failed', 'Failed')],
+        compute='_compute_job',
+        help="Latest Job Status",
+    )
+    job_info = fields.Text(
+        compute='_compute_job',
+        help="Latest Job Error Message",
+    )
+    file_ids = fields.Many2many(
+        comodel_name='ir.attachment',
+        compute='_compute_file',
+        help="List all files created by this report background process",
+    )
+
+    @api.multi
+    def _compute_job(self):
+        for rec in self:
+            rec.job_ids = self.sudo().env['queue.job'].search(
+                [('func_string', 'like', 'report.async(%s,)' % rec.id),
+                 ('user_id', '=', self._uid)],
+                order='id desc')
+            rec.job_status = (rec.job_ids[0].sudo().state
+                              if rec.job_ids else False)
+            rec.job_info = (rec.job_ids[0].sudo().exc_info
+                            if rec.job_ids else False)
+
+    @api.multi
+    def _compute_file(self):
+        for rec in self:
+            rec.file_ids = self.env['ir.attachment'].search(
```

Comment:
Move the search before the loop with ('res_id', 'in', self.ids) and then filter.

It can avoid many sql queries.

Reply:
Great idea. Fixed.